### ISSUE TITLE
Preview feature feedback: Views

### ISSUE THREAD

**Issue by floelhoeffel** [prisma/prisma#17335]
Views support is in early preview and you can learn about it in our [documentation](https://prisma.io/docs/concepts/components/prisma-schema/views).

Please share your feedback about the `views` functionality released in [v4.9.0](https://github.com/prisma/prisma/releases/tag/4.9.0) in this issue.

- If you encounter a bug, please [open a bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=bug_report.md&title=) in this repo.
- If the feature is working well for you, please share this in a comment below or leave a 👍 on this issue.
- If you have any questions, don't hesitate to ask them in the [Prisma Slack](https://slack.prisma.io/) community

**Comment by Ustice** [comment]
The general process is that we add the SQL to generate the views by hand to our migrations, and this new feature gives us a way to describe those views, and query them with the prisma client.

Is this correct?

**Comment by floelhoeffel** [comment]
@Ustice Exactly - the next step is we add introspection support.

**Comment by jonschlinkert** [comment]
I love this feature, but I have some questions. How do views determine field associations? In the following, how would the `UserInfo` view know which table the `type` field belongs to, and how does `UserInfo` know that it's associated with `User`? 

```prisma
view UserInfo {
  id    Int    @id
  name  String
  type  String
}

model User {
  id      Int      @id @default(autoincrement())
  name    String
  type    String?
  profile Profile?
}

model Profile {
  id     Int    @id @default(autoincrement())
  type   String
  user   User   @relation(fields: [userId], references: [id])
  userId Int    @unique
}
```

**Comment by janpio** [comment]
@jonschlinkert That happens via the underlying SQL query that is used to create a view `UserInfo` that has three fields `id`, `name` and `type`? See this documentation section for an example for exactly this model: https://www.prisma.io/docs/concepts/components/prisma-schema/views#how-to-add-views-to-your-prisma-schema The `view` block just tells Prisma which fields to build from the columns the view returns.

Right now you can not define this SQL query in Prisma schema at all to migrate it (via `migrate dev` or `db push` and have to create this manually, outside of Prisma. In the future that might change. 

But, we are already pretty certain that we will not be able to represent these queries with Prisma schema syntax, so for example a "mapping" or "field associations between `UserInfo` and `User`". A view can be defined via _any_ SQL query - which includes an infinite number of possibilities, so we found no way to reliably represent that in Prisma schema.

Does that explain the current state and where this might go?

**Comment by sixers** [comment]
❤️ I love this feature. I was able to move complex multi-table pricing logic, which had to be repeated multiple times in the code to a centralized view. What I liked the most is that I can still model relationships and they just work - which means I can create views which only hold custom computed values, and join them with the original model to get the non-computed fields.

The only annoyance I had so far is the requirement to have an @id field, which doesn't make sense for my table, but I had to create a fake column anyway.

I don't expect this feature to be super popular because it requires SQL knowledge and is probably only beneficial for more complex table structures, but for me it is very important and simplifies a lot, so thank you!

**Comment by janpio** [comment]
Would you have been able to easily identify any of the columns as `@unique` in your view? It will be quite some while until we can fully get rid of the limitation of needing a field to uniquely identify a row, but maybe `@unique` is the better choice than `@id`.

**Comment by sixers** [comment]
My view is a cross join between 2 tables, so I could use compound key (id from first table, id from second table), but not a single column.

**Comment by janpio** [comment]
You would also be able to use `@@unique([first_id, second_id])` as a compound unique if you include both fields in the view. (Sorry, missed to mention that above - we would of course support both `@unique` for individual columns but also `@@unique` for when columns need to be combined.)

**Comment by janpio** [comment]
@Ustice @jonschlinkert If you are using PostgreSQL, have a look at this comment I just posted at the feature request issue: https://github.com/prisma/prisma/issues/678#issuecomment-1412344738

**Comment by dchenk** [comment]
I've found **materialized** views highly useful in the past, and I'm working on an app right now where I'd have loved to have materialized views (so that I can have indexes on a couple columns). What's the timeline for materialized views?

**Comment by janpio** [comment]
We do not communicate concrete timelines. You can already create a materialized view today (just create one instead of a "normal" view) and represent it with the `view` keyword, and then call the "refresh" command via our raw query APIs. That should work. (Would be great if you could try it and give feedback!) When we add materialized view support, we will add an explicit Client API for the refresh command.

**Comment by 5hee75** [comment]
I think this feature, combined with the multi-schema support, is going to be a huge benefit for us poor souls who have been trying to use Prisma with Supabase!  I might try to hack something together this weekend and report back the results.

**Comment by nmfrankel** [comment]
> @jonschlinkert That happens via the underlying SQL query that is used to create a view `UserInfo` that has three fields `id`, `name` and `type`? See this documentation section for an example for exactly this model: https://www.prisma.io/docs/concepts/components/prisma-schema/views#how-to-add-views-to-your-prisma-schema The `view` block just tells Prisma which fields to build from the columns the view returns.
> 
> Right now you can not define this SQL query in Prisma schema at all to migrate it (via `migrate dev` or `db push` and have to create this manually, outside of Prisma. In the future that might change.
> 
> But, we are already pretty certain that we will not be able to represent these queries with Prisma schema syntax, so for example a "mapping" or "field associations between `UserInfo` and `User`". A view can be defined via _any_ SQL query - which includes an infinite number of possibilities, so we found no way to reliably represent that in Prisma schema.
> 
> Does that explain the current state and where this might go?

<br>

Offer the option to reference a column by the table name as an optional alias, for example

```prisma
view UserInfo {
  id    Int    @id
  name  String
  User.type  String
}
```

**Comment by psiho** [comment]
This is a great start! I love it. Comming from SQL world, I'm using Prisma to speed up development (and type safety), not to avoid SQL. It works extremely well for simple stuff which covers 90% of cases. But complex, nested and aggregated queries were a pain to write in Prisma. Even more pain to realize how many individual queries Prisma creates for those. So raw queries and manual types was the only way out. Now this seems to be the best from both worlds and saves a ton of time!

**Comment by luke-cbs** [comment]
Quick question do not know how feasible this is.

So one thing is noted that will be difficult to write the SQL from the Prisma schema itself. Would it maybe be possibly if we have to go down the create-only flag path that maybe we are able to put the SQL scripts files, that create the views themselves with the our Prisma schema and just tell the prisma schema which script file it need to execute to create it? Just to avoid having to do this each time.

```
view UserInfo {
  script = "./UserInfoView.sql" 
  id    Int    @unique
  email String
  name  String
  bio   String
}
```
or maybe if a view is made up of one or more views.
```
view UserInfo {
  script = ["./UserInfoView.sql"]
  id    Int    @unique
  email String
  name  String
  bio   String
}
```
This would just mean it keeps this central. Unless I missed something.

Just a thought.

**Comment by janpio** [comment]
Yes, something in that direction is also our current thoughts re how to support creation and migration of views @luke-cbs. We can not "generate" the SQL from the `view` definition as we can for `model` as the `SELECT` query can be _anything_. Putting the SQL in a file still is challenging though, as the SQL the database stores is slightly different what you put in, and we need to be able to `diff` that to see if we need to update the view and so on - but that is what we will look at after most of the other things work for views. (No ETA, but definitely as one of the later things in this project.)

**Comment by luke-cbs** [comment]
> Yes, something in that direction is also our current thoughts re how to support creation and migration of views @luke-cbs. We can not "generate" the SQL from the `view` definition as we can for `model` as the `SELECT` query can be _anything_. Putting the SQL in a file still is challenging though, as the SQL the database stores is slightly different what you put in, and we need to be able to `diff` that to see if we need to update the view and so on - but that is what we will look at after most of the other things work for views. (No ETA, but definitely as one of the later things in this project.)

Yeah I think there will have to be some give and take from the dev's point of view as you have eluded to. Our responsibility will be by the sounds of it to ensure that our select from these views that we create matches what we define in the .schema.

I think for the most part the diff would be nice. I think even a rough point of re-creating the views when we doing migrations and stuff so that we don't to think why something may be missing or creating some form of area as a source of truth of what the db requires.

But yeah sounds good to me! Thanks for this.

**Comment by luke-cbs** [comment]
One other query I have with support in Webstorm. As currently using the plugin there. Not sure when support will be added for this?

![image](https://user-images.githubusercontent.com/57114092/217917884-33e4d0ad-e079-4e0e-b5c2-f383cc43afa2.png)

When run they work perfect. Just complains.

**Comment by janpio** [comment]
Seems that https://plugins.jetbrains.com/plugin/20686-prisma-orm has not been updated with the latest features yet. Source code lives here: https://github.com/JetBrains/intellij-plugins/tree/master/prisma if you maybe want to help out.

**Comment by luke-cbs** [comment]
> Seems that https://plugins.jetbrains.com/plugin/20686-prisma-orm has not been updated with the latest features yet. Source code lives here: https://github.com/JetBrains/intellij-plugins/tree/master/prisma if you maybe want to help out.
> 
> Our own VS Code extension is fully updated.

Fair! I actually thought that was managed by Prisma not Intelli!

**Comment by bam4564** [comment]
Hey everyone! Love this feature, but have a quick question related to evolving views / their underlying tables via migrations. I'll give a quick example to show my issue. I have the following `Order` model in my database 
```prisma
model Order {
    ...
    priceToSpot Decimal 
    ...
}
```
And additionally, an existing view that computes some properties for the order 
```prisma
view OrderInfo {
  orderId         Int         @id
  order           Order? 
  filledAmount    Decimal     @db.Decimal(78, 0)
  remainingAmount Decimal     @db.Decimal(78, 0)
  status          OrderStatus
}
```
I set the existing model + view up using the docs on views, and everything is working fine, but now I need to change the underlying `Order` model. The `priceToSpot` field is being renamed to `discountRate` and it's semantics are different. In addition, I want to add a new computed field to `OrderInfo` that references the `discountRate` 

It's not clear to me how to perform migrations to get my database to a new stable state. I don't think I'd want to go edit the manual migration I previously created to create the view, as that could have downstream impacts. Could I get a little feedback on how I might accomplish this?

**Comment by sixers** [comment]
@bam4564 I think you can run `prisma migrate dev --create-only`, and then manually edit new migration file to update your view. See https://www.prisma.io/docs/concepts/components/prisma-migrate/migrate-development-production#customizing-migrations

**Comment by bam4564** [comment]
@sixers I'm aware of that, however this is the issue I see. 

To have my updated view `OrderInfo`, it needs to reference the `discountRate` field that I'm adding into the `Order` model. So this would suggest `Order` needs to be migrated first. But wouldn't this create an issue since if I migrate `Order`, dropping `priceToSpot` and adding `discountRate`, that won't be allowed bc my existing view query references `priceToSpot`. 

And I obviously can't update the view before I update the `Order` model with the new column. 

It's kinda a chicken before the egg problem

**Comment by bam4564** [comment]
Actually, I think I might have a workaround here. Even though my `Order` model has a relation to `OrderInfo` view, when looking into the migration, it doesn't actually seem like this relationship is created at the db level, just the prisma client level for querying purposes. So could I do something like this. 
1. Migrate `Order`, dropping `priceToSpot` and adding `discountRate`. 
2. manual migration to modify `OrderInfo` view. Use `CREATE OR REPLACE VIEW` to update the query with my new logic. 

Does this seem like a workable approach here? Documenting just in case this is correct / others find it useful.

**Comment by janpio** [comment]
I think you would indeed use `prisma migrate dev --create-only`:
1. Start by updating the `view` and `model` in your Prisma schema. 
2. Then run the CLI command to get the SQL that would be required to modify the `Order` model and hence `Order` table. 
3. The `view` change will not create any SQL to run, so if you need to update the view via SQL you can add that. If it is just a `SELECT * FROM Order` like select query, you might not even need to. 
4. Then you run the migration with `migrate deploy` and the SQL will be executed. 
5. (If you are on PostgreSQL you can then also run `prisma db pull` to make sure that the new SQL of the view indeed matches what you wrote as your `view`.)

Done. Let me know if this works.

**Comment by bam4564** [comment]
@janpio Thanks for the response, this approach is working for me. One small downside here is in cases where I don't actually need to modify the view creation code, I end up having to do this
1. Drop existing views. Actions in step 2 modify columns these views depend on so this needs to happen first. 
2. Migration logic from `prisma migrate dev --create-only` 
3. Re-create views by copy pasting statements from prior manual migrations. 

This does introduce some code duplication in my migration history, but this isn't too bad. Thanks for the solution!

**Comment by janpio** [comment]
Uh, that is an interesting nuance. So you have create a view manually in the past, and now want to get the `view` into your Prisma schema and then make sure that you have a migration file that properly would recreate it in the future? If that is indeed the case, I think we also have a solution for that.

**Comment by rpominov** [comment]
We run into an issue with the `migrate diff` command when using views. 

This command `npx prisma migrate diff --from-schema-datamodel schema.prisma --to-empty --script` creates SQL to drop everything except views. So when we run the SQL we get a error:

```
Error: db error: ERROR: cannot drop table "Build" because other objects depend on it
DETAIL: view "DashboardProject" depends on table "Build"
HINT: Use DROP ... CASCADE to drop the dependent objects too.
```

Prisma version: 4.9.0

**Comment by talkingdonkeyz** [comment]
First of all, I want to express my appreciation for this feature – it's super handy! However, I've run into an issue when fetching deeply nested data from a view. When I try to access files on records on interviews, they appear to be null even though they seemed to be defined when I logged them.

Here is the SQL for the view and its corresponding schema:

`
SELECT
	row_number() OVER () AS id,
	p.id as "presetId",
	i.id AS "interviewId", 
	e.id AS "encodeId",
	r."feedId" as "feedId",
	v.id as "viewId",
	f.id as "fileId",
    COUNT(DISTINCT fv."feedId") as "feedCount",
    COUNT(DISTINCT f."id") as "recordFileCount",
    COUNT(DISTINCT e.id) as "encodeCount",
    CASE 
    	WHEN COUNT(DISTINCT fv."feedId") = COUNT(DISTINCT r.id) THEN true 
        ELSE false 
	END AS "completeInterview"
FROM "Preset" p
LEFT JOIN "Record" r ON p.id = r."presetId"
LEFT JOIN "File" f ON r."fileId" = f.id
LEFT JOIN "View" v ON v."presetId" = p.id
LEFT JOIN "FeedOnView" fv ON fv."viewId" = v.id
LEFT JOIN "Encode" e ON e."viewId" = v.id
LEFT JOIN "Interview" i ON i."presetId" = p.id
GROUP BY i.id, p.id, e.id, r."feedId", v.id, f.id
`

`
view AggregatedInterview {
  id                Int        @unique
  interviewId       Int?
  presetId          String?    @db.Uuid
  encodeId          String?    @db.Uuid
  feedId            Int?
  viewId            Int?
  fileId            String?    @db.Uuid
  feedCount         Int
  recordFileCount   Int
  encodeCount       Int
  completeInterview Boolean
  interview         Interview? @relation(fields: [interviewId], references: [id])
  preset            Preset?    @relation(fields: [presetId], references: [id])
  encode            Encode?    @relation(fields: [encodeId], references: [id])
  feed              Feed?      @relation(fields: [feedId], references: [id])
  view              View?      @relation(fields: [viewId], references: [id])
  file              File?      @relation(fields: [fileId], references: [id])
}
`

And here is the query I'm using:

`
const readyForEncode = await prisma.aggregatedInterview.findMany({
        where: {
            completeInterview: true,
            encodeId: null,
        },
        include: {
            interview: {
                include: {
                    records: {
                        include: {
                            file: true
                        }
                    }
                }
            },
            view: {
                include: {
                    preset: true,
                    artworkOnView: {
                        include: {
                            artwork: {
                                include: {
                                    file: true
                                }
                            }
                        }
                    },
                    feedOnView: {
                        include: {
                            feed: true
                        }
                    },
                    solids: true
                }
            }
        }
    })
`

I'm not exactly sure why this is happening. It's possible that there's an issue with the SQL query or schema definition, or there may be some issue with the way I'm using Prisma.

If anyone has any insight into what might be causing this issue, or any suggestions for how to fix it, I would greatly appreciate it. Thank you!

**Comment by janpio** [comment]
Can you open a new issue @talkingdonkeyz and provide a bit more information, like the other models that are references in your code so we can try to reproduce this? Happy to dive into this.

**Comment by flaviovich** [comment]
Hello dear!
I'm getting this error using views:

> Cannot read properties of undefined (reading 'findMany')

On local machine it works fine.
I think I'm missing something when I deploy my api.

**Comment by janpio** [comment]
I think that is unrelated to the preview feature @flaviovich - best open a new discussion for this and we can debug what might be going on there.

**Comment by janpio** [comment]
Hey, with 4.12.0 we release the next phase of our `view` preview feature: 

When you use `npx prisma db pull` with the preview feature `views` added to your schema, it will now introspect the SQL of your views and output them into `.sql` files in a new `views` folder right next to your Prisma Schema file. They are grouped into subfolders by the schema name, so this should also work with the `multiSchema` preview feature enabled.

Note: These .sql files are not yet used for creating or updating views during migrations yet!

But we are looking for feedback. Let us know if the introspected SQL files match the views picked up in your database and if the correct files were created in your filesystem. Thanks!

**Comment by atjourneytech** [comment]
Great feature @janpio!

I noticed in a PostGIS database running `npx prisma db pull` it creates .sql views for `geography_columns` and `geometry_columns`.

Not really a problem since it is automatically added with `@@ignore`, however it would be great to have Prisma ignore it altogether. I'd rather not having postgis views in my `prisma.schema`.

I found this: https://github.com/prisma/prisma-engines/blob/4.12.0/migration-engine/connectors/sql-migration-connector/src/sql_schema_differ/sql_schema_differ_flavour/postgres.rs#L26

So maybe there's an issue with the prisma introspection? Thanks.

**Comment by shellscape** [comment]
@janpio The way that the views feature is currently implemented makes automation of `db pull` and `db generate` extremely difficult. Namely the fact that some manual step has to be performed between the two of them to remove the optional `?` and inject an appropriate `@unique` or `@id`. Your competitors use metadata files or db comments to control this - which are optimistic rather than passive.

A core tenet of our CQRS implementation is a schema for read ops, mostly populated by views which controls the public API, and a schema for write ops from which the public API projects nothing. I was hopeful for decent support, but am disappointed by the lack of an optimistic mechanism for deriving relationships.

**Comment by janpio** [comment]
That indeed looks incorrect @atjourneytech, we'll look into that. Thanks for the mention!

We are aware of the current limitations @shellscape - that is why it is still a preview feature 😄 
Unfortunately, the databases do not give us the necessary information for optionality and uniqueness for columns in views, so for now we have to rely on these manual steps. But _of course_ we want to get rid of them, if we can only figure out a way to do that reliably. (We _could_ try to go to the underlying table to get the optionality and uniqueness - but that requires parsing the SQL query of the view to get the table name, and then still make a bunch of assumptions, which _might_ work often - but definitely not always.). 
Do you have examples for these "metadata files or db comments" that other ORMs seems to be using? That might be an interesting lead.

**Comment by shellscape** [comment]
> Do you have examples for these "metadata files or db comments" that other ORMs seems to be using? That might be an interesting lead I am happy to look into.

https://www.graphile.org/postgraphile/smart-tags/ - using `COMMON ON` to add comments to database entities, mysql and postgres only afaik.

https://www.graphile.org/postgraphile/smart-tags-file/ - We've typically used yaml and converted to json with Postgraphile

The latter example follows Prisma's schema language fairly closely. Putting these kinds of assignments in the prisma schema first, and then having Prisma use those assignments during `db pull` to populate the rest of that model's properties would be a reasonable route to go. At the end of the day, I want to define these things ahead of `db pull` and not after, so that we can maximize automation.

**Comment by shellscape** [comment]
OK thanks, I'll give that a shot and report back (may be a week or more)

**Comment by janpio** [comment]
What database? What does the view SQL look like for `PersonComputed`? Or even better, all SQL for both the table(s) and view?

Generally, relations for views for now just work accidentally - we have not really figured out how they should or could work long term.

**Comment by shellscape** [comment]
@janpio testing has gone well predefining keys and relationships so far. do you have any updates on conversations around the information I provided in https://github.com/prisma/prisma/issues/17335#issuecomment-1494610077 ?

**Comment by nkovacic** [comment]
@janpio we are testing `prisma db pull` and it seems it does not support materialized views in PostgreSQL (`views` folders is empty)? 
We are using version `4.12.0`.

**Comment by janpio** [comment]
Good point @nkovacic, we currently do not introspect _materialized_ views at all yet, I opened an issue to make that explicit: https://github.com/prisma/prisma/issues/18758 We'll add that a bit later, but it will be part of the full views support that we are working on.

**Comment by pmosconi** [comment]
Hi I upgraded prisma to `4.12.0` but I don't see view folder and the sql file: I tried `prisma db pull` both for MySql and SqlServer databases. Am I missing something?

**Comment by pimeys** [comment]
> Hi I upgraded prisma to `4.12.0` but I don't see view folder and the sql file: I tried `prisma db pull` both for MySql and SqlServer databases. Am I missing something?

Are you having the preview feature enabled. Did you get the `view` block to the schema file?

**Comment by muzzsteak** [comment]
Hello, I added this as an issue, but perhaps just a comment is more appropriate. 

I am trying to run this migration but I get the error below. Is security_invoker not supported? Or do I perhaps have the syntax wrong?

```
ALTER VIEW IF EXISTS view_xxxx
SET (security_invoker = true);
```

Error: P3006

Migration 20230522152040_alter_view_xxxx failed to apply cleanly to the shadow database.
Error:
db error: ERROR: unrecognized parameter "security_invoker"
0: sql_migration_connector::validate_migrations
with namespaces=None
at migration-engine\connectors\sql-migration-connector\src\lib.rs:301
1: migration_core::state::DevDiagnostic
at migration-engine\core\src\state.rs:269

**Comment by rwieruch** [comment]
Not sure if I understood the concept of views correctly (see [discussion](https://github.com/prisma/prisma/discussions/20290)), but if I am right this would be a great addition to Prisma! Thanks for working on it!

**Comment by aruns05** [comment]
Hello @floelhoeffel Fred
We are not able to import the view as we cannot define any unique constraint 
Can you advise on that ?

**Comment by Jolg42** [comment]
@aruns05 So if you can't add an `@unique`, alternatively you can add an `@id` or `@@id([field1, field2])` or `@@unique([field1, field2])`.

**Comment by dfoley-greenlight** [comment]
@janpio 

Predefining 

> If you are open to modifying your Prisma schema file manually before running `db pull`, you can just put a skeleton `view` in there that already includes `@(@)id`, `@(@)unique` and the fields with types and `?` - or not. The optionality and constraint information will all stick around after you run `db pull` (as we know we can not get that information from the database).

When i do this, "db pull" reverts the fields in my composite id (i.e. expressed via @@id([...]) back to optional, breaking the constraint that the fields must be required. An initial manual edit of my .schema to fix up the view definition is fine, but I can't keep deleting those optional flags ("?") every time i run "db pull". Am I missing something @janpio?

**Comment by janpio** [comment]
Can you share a before and after example? I have trouble imagining the details of what you describe.

**Comment by dfoley-greenlight** [comment]
> Can you share a before and after example? I have trouble imagining the details of what you describe.

Sure. Here are some steps describing the issue, with attached outputs:

1. Run "db pull" to create the initial view definition in the schema.prisma file - this works as per your documentation (File: Initial.txt)
2. Edit the initial view definition created in step 1 as per your documentation - remove "@@ignore", create a composite unique and/or id tag via "@@unique"/"@@id", make fields in composite key required (i.e. delete "?" for those fields). (File: "Edited.txt")
3. Run db-pull again at some point in the future. Result: Composite fields are overridden with their optional flags (i.e. "?") again. (File: 2nd Db Pull.txt)
[Initial.txt](https://github.com/prisma/prisma/files/12514956/Initial.txt)
[Edited.txt](https://github.com/prisma/prisma/files/12515004/Edited.txt)
[2nd Db Pull.txt](https://github.com/prisma/prisma/files/12515005/2nd.Db.Pull.txt)

Note that then 2nd "db pull" doesn't _completely_ override the view definition as the "@@unique" tag persists.

Sql Server
Prisma 5.1

**Comment by eugmakhnev** [comment]
Hello! Thank you for the cool feature and for Prisma itself :)

Do you have plans to support some kind of client-side views? The main idea is to minimize all the hassle with migrations for any kind of read-only tasks, such as views. It seems like a really cheap improvement; you only need to provide a method to replace the view name with a raw query. This could be specified directly in the Prisma client configuration code.

**Comment by janpio** [comment]
That has not been requested yet, @eugmakhnev, so we have not considered or planned that yet. Please open a new feature request issue!

**Comment by kyle-turn2** [comment]
I am able to `db pull` from a `sqlserver` that only has views available. I get the `schema.prisma` and add a `@id` and `@unique` on the appropriate field but am getting a `TypeError: Cannot read properties of undefined (reading 'findMany')` when trying to use it like this:
```
    const employees = await prisma.employeeData.findMany({
      skip: 0,
      take: 10,
    });

    return { employees };
```

This is what one of the original views looked like after I ran `db pull`:
```
/// The underlying view does not contain a valid unique identifier and can therefore currently not be handled by Prisma Client.
view EmployeeData {
  Organization_ID Int?
  Project_ID      Int
  lastname        String    @db.NVarChar(100)
  firstname       String    @db.NVarChar(100)
  ID              Int       
  EmployeeCode    String    @db.NVarChar(25)
  BirthDate       DateTime? @db.Date
  SSN             String?   @db.NVarChar(20)

  @@ignore
}
```
and this is how it looks now:
```
view EmployeeData {
  Organization_ID Int?
  Project_ID      Int
  lastname        String    @db.NVarChar(100)
  firstname       String    @db.NVarChar(100)
  ID              Int       @id @unique
  EmployeeCode    String    @db.NVarChar(25)
  BirthDate       DateTime? @db.Date
  SSN             String?   @db.NVarChar(20)
}
```

Any ideas?

**Comment by ajmnz** [comment]
👋

Loving views so far! Been using them in prod since they launched, and the increased type-safety + being able to define relations (and query them as such) has been delightful. Some feedback after months of using them:

**Relations are great, but they don't really reflect the database schema**

Again, being able to define relations within views and query them like regular tables is truly awesome and has helped fill some of the gaps Prisma has when writing somewhat complex queries. However, the way they are defined in the Schema doesn't match what's really happening in the database (at least on Postgres).

For example, see how `User.powerUser` looks like a regular model relation when in reality is just a "virtual" relation, since it's pointing to a view. These don't exist in the database, so maybe a keyword for clarification can be required (ie. `@view`)

```prisma
model User {
  id        Int       @id @default(autoincrement())
  posts     Post[]
  powerUser TopUser[]
}

model Post {
  id       Int    @id @default(autoincrement())
  user     User   @relation(fields: [userId], references: [id])
  userId   Int
}

view TopUser {
  user     User   @relation(fields: [userId], references: [id])
  userId   Int    @unique
  points   Int
}
``` 

**Migrations are quite difficult and potentially destructive**

Schema changes to a view do not apply to migrations and one must generate the migration manually via `--create-only`, which, don't get me wrong, is alright.

We've seen some friction on fast-changing repositories with lots of migrations, often by different people, where one can get lost when trying to update an existing view. We must check migration by migration in order to find the last to modify the view, then copy it and paste it in a new migration, which can lead to someone copying the wrong version and messing something up — the damage probably won't be _that_ significant, but can happen.

Prisma could (A) expose a CLI option to create a migration for a view based on the latest CREATE state of that view, or (B) ideally intercept changes in the view within the schema and pre-fill the migration with the latest CREATE state of the view. Not sure if the CREATE statement can be pulled from the database, probably not, so the best way could be to just find the latest migration that touched that view.

Hope it helps 🙌🏼

**Comment by jlkravitz** [comment]
I don't really have enough data to report an issue, but I run into issues on Prisma with SQL Server when I use views. I get data conversion errors, e.g.:

```
Message: Invalid `prisma.scores.findMany()` invocation: Error occurred during query execution: ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(TokenError { code: 245, state: 1, class: 16, message: "Conversion failed when converting the varchar value '0cfc75b54b0ada70960afe23162fb674' to data type int.", server: "xxxx", procedure: "", line: 1 }), transient: false })
```


I'm able to fix it by deleting the view in my database, and then recreating it. This happens consistently when I make any changes to the underlying table that the view queries.

**Comment by majimaccho** [comment]
I love this feature. Thank you for your work!

I currently struggling to find a way to distinguish table and view with an internal `getDMMF` method.
Since Redwood.js, which I'm currently using, handles data deletion in each api test with models retrieved from `getDMMF`.

If you can provide any hint like `isView`, I can fix view handling inside Redwood.js.

ref: https://github.com/redwoodjs/redwood/blob/main/packages/testing/config/jest/api/jest.setup.js#L136

**Comment by danbord** [comment]
works great for me, thanks

**Comment by dawid-bobowski** [comment]
Hello!

I've been testing this feature in my setup for the first time today and one question came to my mind regarding the views directory which is created after `npx prisma db pull`.
It creates a new subdirectory named after the database schema. In my case I wouldn't like it to be named this way, because of how my database provider enforces naming of my schemas. They use the name of the main MySQL account as a prefix and I wouldn't like to keep it exposed in my repository.

Is there a way to make this subdirectory's name custom? If not, perhaps some custom folder naming and then using some environment variable to insert the name of the schema would be useful.

Thanks!

**Comment by majimaccho** [comment]
Hi. Thank you for this great feature. I really appreciate this ❤️ 

I found an issue with one to many relation between model and view.

Currently prisma.schema requires FK is defined for one to many relation, however, views shouldn't have any FK since its underlying model has it.

So I suppose the rule to enforce FK to one to many relation would be better to be ignored for relation between view and model.

Thank you!

**Comment by Brokemia** [comment]
I think I'm here with the same concern as @majimaccho 

My case is something like this:
```
model Experiment {
  id          Int         @id @default(autoincrement())
  ...
  conditions  Condition[] 
}

view ExperimentWithOtherData {
  id          Int         @id
  ...
  conditions  Condition[] 
}

model Condition {
  id           Int        @id @default(autoincrement())
  experiment   Experiment @relation(fields: [experimentId], references: [id], onDelete: Cascade)
  experimentId Int
  ...
}
```
Where I'd like Experiment and ExperimentWithOtherData to both have the same list of conditions, since their id fields are the same. Currently I don't see a clean way to do that.

**Comment by mayanafit** [comment]
Hello, has someone faced with issue when deleting the original row from the table, but it's throwing error saying "can't delete from join view" from the view? I've checked it directly from the MySQL, the updatable is true for the view.

**Comment by gustavo-m-vieira** [comment]
Hey! I am trying to use with materialized view at Postgresql but I am having some issues. Right after I deployed my materialized view, when I try to create a new migration or even reset the database I receive the following error:

`The fallback method for database resets failed, meaning Migrate could not clean up the database entirely. Original error: 
ERROR: cannot drop table "Bills" because other objects depend on it
DETAIL: materialized view "CurrentClientBill" depends on table "Bills"
HINT: Use DROP ... CASCADE to drop the dependent objects too.
`
(**CurrentClientBill** is my materialized view and **Bills** is the table in the select to create the view)

Looks like the reset command should use **CASCADE** or the view should be consider at the reset process, being deleted before any of its relations.

There is something that I am missing here?

**Comment by liptonbeer** [comment]
Hi Everyone,
In my project we really like this feature and would like to use it in Production environment. It's been almost 2 years when this feature request has been created and still it is in Early Preview state. Do you have any plans to push it to GA?

**Comment by wmadden** [comment]
A quick update, we plan to release this preview feature to GA between March - August 2025. Read more here: https://github.com/prisma/prisma/discussions/26136

**Comment by 15baraniana** [comment]
Thanks for developing this feature! One thing that doesn't seem to work currently is @map on the fields or @@map on the view name.

In Postgres we keep everything in snake case and then use the Prisma schema to map it to camelCase. I would love to keep that consistent for views as well but it doesn't look like the mapping works. Let me know if this is doable or planned. Thanks!

**Comment by martinkupa** [comment]
It would be really helpful if we could opt out of opposite relations when making views. In my use case the opposing model already has a relation with the underlying table the view represents and adding a repeat field for the view clutters the model.

**Comment by davidcarboni** [comment]
Hi there, I'm really pleased to be picking up Prisma as this feature is nearing GA, it looks like it's going to be a key one for the project I'm working on.

I'm using Postgres with `previewFeatures = ["multiSchema", "views"]` (`^6.3.1`) and noticed that I don't seem to be able to introspect a view I've created in the database, and it also doesn't seem to be generated in the client, even if I add it manually. I've added `@id` and `@unique`.

If could be my mistake so I'm going over the documentation again but I thought I'd mention this in case it's a known issue?

**Comment by EDM115** [comment]
heyy :)  
any news or ETA for materialized views ?

**Comment by davidcarboni** [comment]
I've tried to add views to my migrations, but I'm getting an error:

```
Error: P3006

Migration `20250314182241_xxxx` failed to apply cleanly to the shadow database. 
Error code: P1014
Error:
The underlying table for model `grouse.LicenceApplication` does not exist.
```

I can run `npx prisma migrate reset` cleanly, but if I ten try to run `npx prisma migrate dev --name xxxx` I get this error.

I noticed if I comment out the SQL in the view migrations I'm able to create a migration.

Steps to get here:
 * Create a model in the normal way (standard, works fine)
 * Create a view manually in the database (actually, 2 views, same process for both)
 * `npx prisma db pull` to add the view to `schema.prisma` and update according to `https://www.prisma.io/docs/orm/prisma-schema/data-model/views`
 * I now have `prisma\views\xxxx.sql` which comtains the body of the view, but not the `create view ...` part.
 * I run `npx prisma migrate dev --create-only --name xxxx` and paste in the sql, plus `create view ...`
 * Add a new model `yyyy` to `schema.prisma`
 * Run `npx prisma migrate dev --name yyyy` -> `Error: P3006`
 * Comment out sql of the views, now I can create a migration

**Comment by GabrielRMCorrea** [comment]
Hello,
I noticed that in Prisma.dmmf.datamodel, views are treated equivalently to models. I couldn’t distinguish between a view and a table when working with it. Having a different category for views would help me a lot

**Comment by velipso** [comment]
I'm using v5.7.0. I created a materialized view (PostgreSQL) manually in a migration... now when I `prisma migrate dev --create-only`, it always generates a migration file, but the file only contains `-- This is an empty migration.`

It seems the migration tool is never satisfied and always thinks there needs to be one generated, but then realizes late that there is nothing to do.

Not sure if this is a bug with my version or not, but figured I would post it.

**Comment by GauBen** [comment]
Hi there, the latest Prisma release (6.13.0) broke the ability to use relations with views:

```
Error validating: The field `author` is part of a relation that references a view. Relations currently cannot be used with views.
```

This is a hard blocker on my end and it prevents upgrading (views are useful because I'm able to link them to models). I hope we have a way to declare relations between views and models for the GA of the feature.

**Comment by aqrln** [comment]
@GauBen we likely won't. It is exactly due to us moving the `views` preview feature closer to GA and stabilizing its API that we have decided to remove support for relationships involving views after careful consideration, in addition to some other Client API restrictions. In exchange, we have lifted the requirement to have a unique field in the view. In fact, you can't mark any fields inside the view blocks as `@id` or `@unique` anymore (`@unique` may be supported for materialized views in the future).

Relationship support for views in the way we provided was inherently unsafe and had footguns we saw users constantly running into, causing correctness issues in their applications. Some of those problems were unsolvable without limiting other aspects of the views support. It is possible to design and re-introduce a safe, limited subset of the former relationship APIs for views though. If you relied on this, please open a feature request and tell us about your use case!

**Comment by sixers** [comment]
It's very unfortunate, my primary use case for views has been to add computed fields to models that involve complex SQL operations and aggregations. I've implemented these as views and then connected them to the main model via relationships. It's a key use case for me, and it seems it's no longer supported.

**Comment by ajmnz** [comment]
Totally agree with @GauBen and @sixers. Relationship support was a key selling point of views for us too. It enabled powerful use cases that are now hard to replicate without it, forcing us to fall back to raw SQL. Would really like to see a safe subset brought back before GA.

**Comment by clemens** [comment]
First of all, to everyone who's having issues with the most recent 6.13.0 release's changes to views, here's a workaround that seems to work for our codebase:
- Enable the `externalTables` preview feature and add the views that use unsupported features to the external tables list
- Change your views to be models (as in: `view SomeView` becomes `model SomeView`)

It's a bit clumsy and the downside is that it adds methods like `insert`, `update` etc. to the models, which obviously won't work. But IMO that's more a developer experience concern than an actual problem: Since you've been using views (and have been using the `view` keyword), you're hopefully not using modifying methods – and if you do, then you have a problem already (with or without the update to 6.13.0).

---

Now a thing for this particular discussion thread.

@aqrln I might be wrong here, but I think these recent changes go against your official [ORM releases and maturity levels](https://www.prisma.io/docs/orm/more/releases). There are two statements in this guide that stand out:

> Preview: [...] Users can count on the feature or product and associated API to be mostly stable unless stated otherwise in the release notes and documentation.

I've done a quick search in previous release notes and couldn't spot any mention of "we're unsure about supporting indexes, constraints and relationships on views, so we might remove them without replacement".

> Breaking changes can still be rolled out in MINOR but only for opt-in Preview and Early Access features that are not active by default (e.g. via a Preview feature flag or a specific opt-in option or new CLI command).

This IMO directly contradicts the above. So at best, it's confusing – and at the worst, one can't know which is true for which case.

This is particularly striking to me, since for the case of views, it seems that this has been [declared as "Preview"](https://github.com/prisma/prisma/releases/tag/4.9.0) for almost 2.5 years (2 major versions ago). Add to that that Prisma literally until 2 weeks ago [enforced adding a unique identifier](https://github.com/prisma/docs/pull/7030/files#diff-3883b8b5d1aaccad5ac9c0e9c90e5755bdd61b5254423abb3fe91a7363a40c0dL261-L293) and [told you it's because it's treated the same as regular models](https://github.com/prisma/docs/pull/7030/files#diff-3883b8b5d1aaccad5ac9c0e9c90e5755bdd61b5254423abb3fe91a7363a40c0dL316-L330) and we end up in the current situation – which is less than optimal. IMO it doesn't fit Prisma's maturity level to instruct people that they need to add a unique identifier to their views, because otherwise it won't work and then do a full 180 on it and tell people that what they've had to do for 2.5 years not only isn't necessary anymore but in fact breaks altogether.

I would encourage the Prisma team to take one of two steps to remedy the situation:
- Roll back the change altogether and introduce an upgrade path; maybe even only in Prisma 7, because even if it is a "Preview" feature, it's arguably a VERY breaking change
- If that's not possible or interferes too much with your overall process and roadmap, maybe you could verify my workaround mentioned at the top of this post – and if it does work, you could consider including it in the [docs for views](https://www.prisma.io/docs/orm/prisma-schema/data-model/views)

If you want to go with option 2 and have verified that it's compatible, I'm happy to submit a PR with better phrasing and examples than above for the docs repo.

**Comment by khalilsarwari** [comment]
We've been using the views preview feature for quite some time now as well, similar use case as mentioned above in which heavy computations are stored and then fetched via relationships. This most recent version invalidates this pretty common and reasonable use case, which it seems many people are reliant upon. 

I can understand the downsides to the original views setup with the footguns, but at the same time it could be leveraged (with care) in a way that did not seem janky and served its purpose well. I would also like to request rolling back the change for now - not only would it improve the impression of stability for prisma, but it would give some more time to think through and potentially incorporate a safe version of the relationship APIs from the get-go to prevent breakage.

**Comment by dvins** [comment]
> [@GauBen](https://github.com/GauBen) we likely won't. It is exactly due to us moving the `views` preview feature closer to GA and stabilizing its API that we have decided to remove support for relationships involving views after careful consideration, in addition to some other Client API restrictions. In exchange, we have lifted the requirement to have a unique field in the view. In fact, you can't mark any fields inside the view blocks as `@id` or `@unique` anymore (`@unique` may be supported for materialized views in the future).
> 
> Relationship support for views in the way we provided was inherently unsafe and had footguns we saw users constantly running into, causing correctness issues in their applications. Some of those problems were unsolvable without limiting other aspects of the views support. It is possible to design and re-introduce a safe, limited subset of the former relationship APIs for views though. If you relied on this, please open a feature request and tell us about your use case!

The silence from the Prisma team on this issue (and the half dozen related ones since opened) is concerning. What's more troubling: the lack of engagement, or that these design decisions suggest a fundamental misunderstanding of how enterprises actually use views in relational databases?

**Comment by videvian** [comment]
> > [@GauBen](https://github.com/GauBen) we likely won't. It is exactly due to us moving the `views` preview feature closer to GA and stabilizing its API that we have decided to remove support for relationships involving views after careful consideration, in addition to some other Client API restrictions. In exchange, we have lifted the requirement to have a unique field in the view. In fact, you can't mark any fields inside the view blocks as `@id` or `@unique` anymore (`@unique` may be supported for materialized views in the future).
> > Relationship support for views in the way we provided was inherently unsafe and had footguns we saw users constantly running into, causing correctness issues in their applications. Some of those problems were unsolvable without limiting other aspects of the views support. It is possible to design and re-introduce a safe, limited subset of the former relationship APIs for views though. If you relied on this, please open a feature request and tell us about your use case!
> 
> The silence from the Prisma team on this issue (and the half dozen related ones since opened) is concerning. What's more troubling: the lack of engagement, or that these design decisions suggest a fundamental misunderstanding of how enterprises actually use views in relational databases?

They reverted some changes in the latest release https://github.com/prisma/prisma/releases/tag/6.14.0
Still no `@id`, but now `@unique` and relations work again.

They did delete some other things in that release like `prisma.$use` and marked `metrics` as deprecated without updating docs, so all examples still show how to use `metrics` with `prisma.$metrics`. But it works for now, so no complaints. Though I guess, they could've deleted `$use` in a major version instead of a minor one.

**Comment by dvins** [comment]
> They reverted some changes in the latest release https://github.com/prisma/prisma/releases/tag/6.14.0 Still no `@id`, but now `@unique` and relations work again.
> 
> They did delete some other things in that release like `prisma.$use` and marked `metrics` as deprecated without updating docs, so all examples still show how to use `metrics` with `prisma.$metrics`. But it works for now, so no complaints. Though I guess, they could've deleted `$use` in a major version instead of a minor one.

That's great news, though bad form nothing communicated on any of the half dozen threads this particular reversion was coming or that it dropped.

**Comment by aqrln** [comment]
Hey folks, I'd like to clear up some misunderstandings and provide an update for you. I was going to do this a while ago but unfortunately this fell through the cracks for which I apologize. We did hear you, and we did re-prioritize work and act on your feedback.

@ethos-vitalii:

> They reverted some changes in the latest release [6.14.0 (release](https://github.com/prisma/prisma/releases/tag/6.14.0)

What we did wasn't just "revert" the changes. We now have a more robust implementation that avoids the pitfalls of the previous implementation, built on top of the work done in 6.13.0 (which, in turn, wasn't just "removing" the functionality, that was more of consequence of the changes than the main change itself). With Prisma 6.14 and newer, views are more of a first class citizen and aren't treated the same as models anymore. This means that:
* Views have their own validations in Prisma schema language and don't trigger model-specific validations that don't apply to them.
* Views don't need to have an ID or unique fields. The available APIs depend on whether there are unique fields. Before Prisma 6.13.0, Prisma forced you to artificially declare at least some field as unique, which led to issues when none of them were actually unique. I understand that those of you who have left feedback in this issue took care to make sure these annotations were correct, but it wasn't the case for many other users who were running into serious correctness issues in their applications because of this. Marking a field as unique is now a conscious opt-in rather than something that ORM forces you to do, and views are now usable in the Client either way, just with different API affordances.

>  Still no `@id`, but now `@unique` and relations work again.

You don't need an `@id`, there was no practical difference between using `@id` or `@unique` in views before 6.13.0. The distinction between them only matters for models, and mostly for Migrate, because they create different kinds of constraints and indexes. Views aren't yet supported by Migrate, and regular (non-materialized) views have no indexes or constraints. `@unique` for views in 6.14.0+ now has clearly defined semantics (and its own validations in PSL that differ from `@unique` for models) of an unchecked assertion that you can use to tell Prisma that you guarantee the uniqueness of the field. We will later support real database-level unique indexes when we add support for materialized views.

@clemens 

> I might be wrong here, but I think these recent changes go against your official [ORM releases and maturity levels](https://www.prisma.io/docs/orm/more/releases)

Thank you, this is a very good callout. I do agree with you that de-facto views feel like an early access feature, given the lack of a stable and validated product surface, especially support in Migrate. I can't comment on whether it satisfied the formal criteria for "Preview" back when the feature was introduced because I don't know. However, you may have noticed that nowadays we use the "Early Access" status consistently and all large features go through this stage, rather than starting in "Preview" immediately. We are committed to bringing our long-standing preview features to GA, and you can be confident in the clarity of the maturity levels of all new features.

**Comment by sixers** [comment]
🙌🏻  Thank you for listening to the community.

**Comment by jdpt0** [comment]
Hi @aqrln, thanks for listening to the feedback. My team, like many others here, would have been severely impacted by this change, so glad to hear you took this on board.

You mention:

> With Prisma 6.14 and newer, views are more of a first-class citizen and aren't treated the same as models anymore.

Does this also mean there will be future updates that allow us to distinguish between a model and a view in the DMMF? This would be very handy, as we're using Prisma Extensions to generate factories for our models. The change to views that removes write support (while itself is a welcome change) breaks this extension and makes it very hard to work around, given that we can't easily distinguish between a view and a model.

**Comment by AjaniBilby** [comment]
It looks like all of the pieces for this feature already exist in Prisma for this to be properly supported (*at least for Postgres servers*), they just need to be connected together.

I just want to confirm if this would be a potential method for support?

If this is a valid approach, if this feature isn't in by Christmas I might be able to get my company to dedicate some resources to it to make a PR to implement the below.
Either that or if someone else sees this and has the motivation they could try?

---

### View Syncing

The bare-minimum should be detecting changes, and automatically creating a new migration.
Since I have already had headaches dealing with database generated columns, it's clear Prisma doesn't currently have a way to syntatically comparing if two SQL statements are identical. But you can get the DB to do the heavy lifting for this.

_Since I haven't looked into how the internals for migration detection work I'm just going to write it as a prisma transaction for demonstration purposes._
```ts
for (const view of views) { // could also be used for detecting db-drift on a shadow db
  try {
    await prisma.$transaction(async (tx) => {
    	// create it as a new view so the database will parse and normalize the syntax
    	const id  = Math.floor(Math.random() * 1679615).toString(36).padStart('0', 4);
    	const tmp = `_prisma_view_tmp${id}`;

     	// compare the database representations of the view (always unmaterialized to avoid excessive computation)
    	await tx.$executeRawUnsafe(`CREATE VIEW "${tmp}" AS ${view.sql}`); // sql should be checked for erroneous semi-columns

     	const [ match ] = await tx.$queryRawUnsafe(`SELECT pg_get_viewdef('"${view.name}"'::regclass) = pg_get_viewdef('"${tmp}"'::regclass) as "match"`);
     	if (!match) view.markOutOfDate();

     	// roll the transaction back to remove the created view (could use drop instead)
     	throw new Error('rollback');
    });
  } catch { /* using errors for rollback */ }
}
```

> Then `views` should be stored in their own folder like `sql`.
> The simplest method would be the file name has to match the view name, or an extra attribute could be added to the view to specify the filename - though I feel like that might be excessive bloat.
> Possibly the lookup for a view's file should be case in-sensitive to deal with window/unix compatibility issues that might otherwise arrise, and ethically a project should have two views with the same name seperated by casing.

---

### Type Validation
Since [typedSql](https://www.prisma.io/typedsql) is already able to resolve the return types of a query, you should be able to piggy back off that to determine the output type of the view, and verify it matches with the schema in the `.prisma`.

At bare minimum you could repurpose parts of the `typedSql` code to at least vaildate the query before attempting to check for migration

---

### Nice to Haves

There should be an `@stored` attribute which could be added to the view itself to enable indexing and then it needs to be created as `MATERALIZED`.
The materialized state can be checked with `SELECT relkind = 'm' as "materialized" FROM pg_class  WHERE relname = ${name};` to determine if the view needs to be created in the correct mode.

---

I personally do not have enough experience with MySQL to comment on it's equivalents for this, but I assume where are some.

---

In the mean time for anyone else here, I have made a [little script](https://gist.github.com/AjaniBilby/9c336c6acfe2465acbeedd2e941d18ad) I will use to help check if my views definition files match what's in my development db, so I can synchonise any view updates easily.
